In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Import the CSV file NSMES1988updated.csv file and create a new dataframe for working with Pandas
data = pd.read_csv('NSMES1988.csv')

data.head()

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


In [3]:
# Identify different types of data and report it.
data.dtypes

Unnamed: 0      int64
visits          int64
nvisits         int64
ovisits         int64
novisits        int64
emergency       int64
hospital        int64
health            str
chronic         int64
adl               str
region            str
age           float64
gender            str
married           str
school          int64
income        float64
employed          str
insurance         str
medicaid          str
dtype: object

In [4]:
numurical_cols = data.select_dtypes(include=['int64','float64']).columns.tolist()
numurical_cols

['Unnamed: 0',
 'visits',
 'nvisits',
 'ovisits',
 'novisits',
 'emergency',
 'hospital',
 'chronic',
 'age',
 'school',
 'income']

In [5]:
# Identify Categorical types in the data.
categorical_cols = data.select_dtypes(include=['object', 'category', 'bool']).columns.tolist()
categorical_cols

/var/folders/f7/6ttbslv14mb1smgbhyrcq1xh0000gn/T/ipykernel_23380/1442199522.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = data.select_dtypes(include=['object', 'category', 'bool']).columns.tolist()


['health',
 'adl',
 'region',
 'gender',
 'married',
 'employed',
 'insurance',
 'medicaid']

In [6]:
data.drop(columns=['Unnamed: 0'], inplace=True)

In [7]:
# Perform a detailed Data pivoting on the dataframe and report it.

data.describe(include='all')

,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
count,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406,4406.000000,4406,4406,4406.000000,4406,4406,4406.000000,4406.000000,4406,4406,4406
unique,NaN,NaN,NaN,NaN,NaN,NaN,3,NaN,2,4,NaN,2,2,NaN,NaN,2,2,2
top,NaN,NaN,NaN,NaN,NaN,NaN,average,NaN,normal,other,NaN,female,yes,NaN,NaN,no,yes,no
freq,NaN,NaN,NaN,NaN,NaN,NaN,3509,NaN,3507,1614,NaN,2628,2406,NaN,NaN,3951,3421,4004
mean,5.774399,1.618021,0.750794,0.536087,0.263504,0.295960,NaN,1.541988,NaN,NaN,7.402406,NaN,NaN,10.290286,2.527132,NaN,NaN,NaN
std,6.759225,5.317056,3.652759,3.879506,0.703659,0.746398,NaN,1.349632,NaN,NaN,0.633405,NaN,NaN,3.738736,2.924648,NaN,NaN,NaN
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,0.000000,NaN,NaN,6.600000,NaN,NaN,0.000000,-1.012500,NaN,NaN,NaN
25%,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,1.000000,NaN,NaN,6.900000,NaN,NaN,8.000000,0.912150,NaN,NaN,NaN
50%,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN,1.000000,NaN,NaN,7.300000,NaN,NaN,11.000000,1.698150,NaN,NaN,NaN
75%,8.000000,1.000000,0.000000,0.000000,0.000000,0.000000,NaN,2.000000,NaN,NaN,7.800000,NaN,NaN,12.000000,3.172850,NaN,NaN,NaN


## Dataset Overview

The dataset contains **4406 records** across 18 variables, with no missing values 
(all columns show count = 4406).

### Numeric Variables

| Variable | Mean | Std Dev | Min | Max | Observation |
|---|---|---|---|---|---|
| visits | 5.77 | 6.76 | 0 | 89 | Highly right-skewed; most patients visit far less than max |
| nvisits | 1.62 | 5.32 | 0 | 104 | Median = 0, so most patients have zero non-physician visits |
| ovisits | 0.75 | 3.65 | 0 | 141 | Rare event with extreme outliers |
| novisits | 0.54 | 3.88 | 0 | 155 | Same pattern — rare but extreme outliers exist |
| emergency | 0.26 | 0.70 | 0 | 12 | Emergency visits are uncommon |
| hospital | 0.30 | 0.75 | 0 | 8 | Most individuals have 0–1 hospital stays |
| chronic | 1.54 | 1.35 | 0 | 8 | Average of ~1–2 chronic conditions per person |
| age | 7.40 | 0.63 | 6.6 | 10.9 | Age recorded as years/10, so range ≈ 66–109 yrs |
| school | 10.29 | 3.74 | 0 | 18 | Average ~10 years of education |
| income | 2.53 | 2.92 | -1.01 | 54.84 | Contains a **negative value** — likely a data-quality issue to flag |

### Categorical Variables

| Variable | Unique Values | Most Frequent (Top) | Frequency |
|---|---|---|---|
| health | 3 | average | 3509 (80% of records) |
| adl | 2 | normal | 3507 (80%) |
| region | 4 | other | 1614 (37%) |
| gender | 2 | female | 2628 (60%) |
| married | 2 | yes | 2406 (55%) |
| insurance | 2 | no | 3951 (90%) |
| employed | 2 | yes | 3421 (78%) |
| medicaid | 2 | no | 4004 (91%) |

### Key Observations
- **No missing data** across any column — dataset is clean in terms of completeness.
- **Income has a negative minimum (-1.01)**, which is not logically valid for income and should be investigated/cleaned.
- **Visit-related columns (visits, nvisits, ovisits, novisits)** are heavily right-skewed — a small number of individuals account for very high visit counts (outliers).
- **Health status** is dominated by "average" (80%), meaning few individuals report poor or excellent health.
- **Insurance and Medicaid coverage are low** (only ~9–10% have Medicaid; most lack private insurance too), suggesting a largely uninsured/self-pay population in this sample.
- **Gender split** is moderately imbalanced (60% female).


In [8]:
# Prepare a detailed report on your analysis and observations.

data[['health', 'region']].info()

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   health  4406 non-null   str  
 1   region  4406 non-null   str  
dtypes: str(2)
memory usage: 69.0 KB


In [9]:
data[['health', 'region']].value_counts()

health     region   
average    other        1237
           midwest       957
           northeast     694
           west          621
poor       other         272
           midwest       110
excellent  other         105
           west           91
           midwest        90
poor       northeast      86
           west           86
excellent  northeast      57
Name: count, dtype: int64

In [10]:
data['region'].unique()

<StringArray>
['other', 'midwest', 'northeast', 'west']
Length: 4, dtype: str

## Analysis of Categorical Columns: Health & Region

### Structure
- The dataframe subset `data[['health', 'region']]` contains **4406 entries** (rows 0–4405), 
  with **no missing values** in either column (both show 4406 non-null).
- Both columns are stored as **string/object dtype**, confirming they are categorical 
  (non-numeric) variables, not requiring conversion.
- Combined memory usage is minimal (~69 KB), confirming this is a lightweight categorical subset.

### Region — Unique Categories
`region` has exactly **4 distinct categories**:
- `other`
- `midwest`
- `northeast`
- `west`

### Observations
- **Completeness:** Both `health` and `region` are fully populated — no cleaning 
  (missing-value handling) is required for these two columns.
- **Health** is categorical with a small number of levels (from earlier `value_counts()`, 
  3 levels: poor/average/excellent), making it suitable for grouping numeric measures 
  like `visits`, `hospital`, `chronic`.
- **Region** has 4 levels, with `other` as a catch-all category rather than a specific 
  named region — worth noting as a limitation, since "other" obscures which real regions 
  it represents.
- Since both columns are clean and fully populated, they are **ready to be used directly 
  as pivot table index/columns** (`pivot_table(index='health', columns='region', ...)`) 
  without additional preprocessing.

### Conclusion
Both `health` and `region` are well-structured categorical variables with no missing 
data, making them reliable grouping keys for the pivot analysis required in this task.


In [11]:
data.to_csv("NSMES1988updated3.csv", index=False)
print("Saved -> NSMES1988updated3.csv")
print("Rows, Columns:", data.shape)

Saved -> NSMES1988updated3.csv
Rows, Columns: (4406, 18)
